File: xy_alignment_script.m
% % % clear
% % % 
% % % %%
% % % nEmbryosUsed = 102 % 102 fly embryos total
% % % inputData = 'gap_data_raw_dorsal_wt'; % variable stored: 'data'
% % % inputFile = [inputData,'.mat'];
% % % 
% % % %%
% % % load(inputFile);
% % % nEmbryos = numel(data);
% % % assert(nEmbryos >= nEmbryosUsed);
% % % 
% % % % Authors' description of the data: "Each profile vector is of length 1000
% % % % pixels, where 0 corresponds to the anterior (A) and 1000 to the posterior
% % % % (P) of the embryo."
% % % % To mitigate 'edge effects,' we extract only the middle 80% of the data to 
% % % % use throughout the analysis.
% % % 
% % % nSamplePts = numel(data(1).Hb);
% % % x = (1/nSamplePts : 1/nSamplePts : 1);
% % % idx_low = round(nSamplePts)*0.1;
% % % idx_high = round(nSamplePts)*0.9;
% % % x = x(idx_low:idx_high);
% % % 
% % % genes = {'Hb','Kr','Gt','Kni'};
% % % nGenes = numel(genes);
% % % 
% % % %% Set up raw data
% % % Y_raw = zeros(nSamplePts,nEmbryosUsed,nGenes);
% % % idcs = sort(randperm(nEmbryos,nEmbryosUsed));
% % % 
% % % for iG = 1:nGenes
% % %     for iE = 1:nEmbryos
% % %         Y_raw(:,iE,iG) = data(iE).(genes{iG});
% % %     end
% % % end
% % % Y_raw = Y_raw(idx_low:idx_high,idcs,:);
% % % [nPts,~,~] = size(Y_raw);
% % % padSize = round(nSamplePts*0.05);
% % % Y_raw = padarray(Y_raw,padSize,NaN,'both');

%% Cochlea data
load('F:\projects\cochlea\data\cochlea-data_2020-07-15_11-36-59.mat')
[fsa,ft] = filtercochleadata(data,12.5,true,{'bmp4_mrna','topro3'});
clear Y_raw
clear Y_yAlign1
clear Y_xAlign
clear Y_align
genes = {'bmp4_mrna'};
nGenes = numel(genes);

gene1 = ft.bmp4_mrna;
gene2 = ft.topro3;

[nSamplePts,nC,~] = size(gene1);

for iC = 1:nC
    gene1(:,iC) = regressnoise(gene2(:,iC), gene1(:,iC));
%     gene2(:,iC) = regressnoise(gene3(:,iC), gene2(:,iC));
    
    Y_raw(:,iC,1) = smoothrawdata(gene1(:,iC),'movmean',50);
%     Y_raw(:,iC,2) = smoothrawdata(gene2(:,iC),'movmean',50);
end

% Y_raw(:,2,:) = [];
% Y_raw(:,5,:) = [];
% Y_raw(:,6,:) = [];
[nSamplePts,nC,~] = size(Y_raw(:,:,1));

% Remove first and last 10%.
nSamplePts = numel(Y_raw(:,1,1));
x = (1/nSamplePts : 1/nSamplePts : 1);
idx_low = round(nSamplePts)*0.1;
idx_high = round(nSamplePts)*0.9;
x = x(idx_low:idx_high);

% Pad ends with NaNs.
Y_raw = Y_raw(idx_low:idx_high,:,:);
[nPts,~,~] = size(Y_raw);
padSize = round(nSamplePts*0.1);
Y_raw = padarray(Y_raw,padSize,NaN,'both');

%% Comments
% Up to this point, we have simply deleted data outside of the middle 80%
% of the domain (replacing it with NaNs to act as a buffer to shift the
% profiles to the left/right during X-alignment).
%% 1 Align Y alone (optimize alpha and beta).
[Y_raw,~] = anchormean0to1(Y_raw);
Y_yAlign1 = zeros(size(Y_raw));
for iG = 1:nGenes
    y = squeeze(Y_raw(:,:,iG));
    [Y_yAlign1(:,:,iG),~] = aligny(y);
end

figure
hold on
plot(Y_raw(:,:,1))
% plot(Y_raw(:,:,2))

%% 2 Numerically estimate optimal joint XY alignment (alpha, beta, gamma).
% UseParallel_TF = false;
% pool = gcp;
% if UseParallel_TF
%     if isempty(gcp('nocreate'))
%         parpool(pool.NumWorkers)
%     end
% end
[Y_xAlign,alpha,beta,gamma] = alignxy(Y_yAlign1,false);
% Should explore a 'better' solution than genetic optimization. This was
% chosen because I shift profiles in x using their integer indices, and
% MATLAB's genetic optimizer allowed for easily implemented integer
% constraints on one parameter (gamma) and continuous variation on others
% (alpha and beta). This approach does not yield the same result twice, but
% reliably yields a better x-alignment than the data came with originally.
% However, it needs tweaking for the data at hand (e.g. the NaN buffer size
% on either side of each profile should be large enough, but should not be
% too large). Additionally, y-alignment is poor when done jointly using
% this method, so an additional chi^2 minimization must be run afterward.

%% Trim away NaNs used for padding and x alignment.
[nX,~,~] = size(Y_xAlign);
Y_xAlign = Y_xAlign(padSize:nX-padSize+1,:,:);

%% 3 Align Y alone once more for closed-form Y alignment.
Y_align = zeros(size(Y_xAlign));
for iG = 1:nGenes
    y = squeeze(Y_xAlign(:,:,iG));
    [Y_align(:,:,iG),~] = aligny(y);
end

figure
hold on
plot(Y_align(:,:,1),'k')
% plot(Y_align(:,:,2),'r')

figure(2)
hold on
plot(mean(Y_align(:,:,2),2),'r','linew',2)

% figure
% hold on
% for iC = 1:nC
%     plot(Y_align(:,iC,1),'b')
%     plot(Y_align(:,iC,2),'r')
%     title(iC)
%     pause(4)
% end

% % % %% Save final result
% % % save([inputData,'_XY-Aligned.mat'])

File: xy_alignment_script.py
import scipy.io
import matplotlib.pyplot as plt
import numpy as np

# Is there a cleaner way to use these functions?
from anchormean0to1 import anchormean0to1
from aligny import aligny,chisq
from alignxy import alignxy#,chisq

inputData = 'gap_data_raw_dorsal_wt' 
inputFile = inputData + '.mat'
mat = scipy.io.loadmat(inputFile)

Hblist = [val for sublist in mat['data']['Hb'][0][0:102] for val in sublist]
Krlist = [val for sublist in mat['data']['Kr'][0][0:102] for val in sublist]
Gtlist = [val for sublist in mat['data']['Gt'][0][0:102] for val in sublist]
Knilist = [val for sublist in mat['data']['Kni'][0][0:102] for val in sublist]

# 3D array with dimensions nSamplePoints x nEmbryos x nGenes
gapDataRawDorsalArray = np.transpose(np.array([Hblist, Krlist, Gtlist, Knilist]))

nGenes=gapDataRawDorsalArray.shape[2] # xy_alignment_script.m:26
nEmbryos= gapDataRawDorsalArray.shape[1] #  = 102
nEmbryosUsed = 102 #102 fly embryos total

assert(nEmbryos >= nEmbryosUsed) # xy_alignment_script.m:11

"""
Authors' description of the data: "Each profile vector is of length 1000
pixels, where 0 corresponds to the anterior (A) and 1000 to the posterior
(P) of the embryo."
To mitigate 'edge effects,' we extract only the middle 80% of the data to 
use throughout the analysis.
"""

nSamplePts = gapDataRawDorsalArray.shape[0] # = 1000
x=(np.arange(1 / nSamplePts,1,1 / nSamplePts)) 
idx_low=np.dot(round(nSamplePts),0.1)
idx_high=np.dot(round(nSamplePts),0.9)
x=np.arange(int(idx_low),int(idx_high))

## Set up raw data xy_alignment_script.m:28

idcs=np.sort(np.random.permutation(np.random.randint(1, nEmbryos,size= nEmbryosUsed)))# xy_alignment_script.m:30
Y_raw = gapDataRawDorsalArray.astype(float) # xy_alignment_script.m:32-36

Y_raw = Y_raw[:,idcs,:]
Y_raw = Y_raw[x,:,:]
padSize=round((nSamplePts*0.1))
Y_raw=np.pad(Y_raw,[(padSize,padSize),(0,0),(0,0)],mode='constant', constant_values=(np.nan))

"""
display anchormean0to1 for Hb gene
"""

x = np.arange(1000)

Y_mean = np.nanmean(Y_raw,axis=1) #"in" plot for Hb
#for i in np.arange(1,102): 
#    plt.plot(gapDataRawDorsalArray[:,i,0],color='grey', linewidth=0.5)
#plt.plot(Y_mean[:,0],color='black')
#plt.title('Python in')
#plt.show()

Y_anchor,Y_mean = anchormean0to1(Y_raw)

#for i in np.arange(1,102): 
#   plt.plot(Y_anchor[:,i,0],color='grey', linewidth=0.5)
   
#plt.plot(Y_mean[:,0],color='black')
#plt.title('Python out')
#plt.show()


# =============================================================================
# ## Initial Y alignment
# Y_yAlign1 = np.zeros(Y_anchor.shape)
# for iG in np.arange(0,nGenes):
#     y = np.squeeze(Y_anchor[:,:,iG])
#     #plt.plot(y,color='grey',linewidth=0.5)
#     #plt.show()
#     
#     y_align_temp = aligny(y)
#     Y_yAlign1[:,:,iG] = y_align_temp
#     
#     #plt.plot(y_align_temp,color='black',linewidth=0.5)
#     #plt.show()
# =============================================================================
    
## XY joint alignment
Y_xyAlign = np.zeros(Y_anchor.shape)

Y_xyAlign = alignxy(Y_anchor) # Pass all genes simultaneously
